libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/entitlements/FilesEntitlement.java

```diff
@@ -29,8 +34,104 @@ public enum Mode {
         READ_WRITE
     }
 
-    public record FileData(String path, Mode mode) {
+    public enum BaseDir {
+        CONFIG,
+        DATA
+    }
+
+    public sealed interface FileData {
+
+        final class AbsolutePathFileData implements FileData {
+            private final Path path;
+            private final Mode mode;
+
+            private AbsolutePathFileData(Path path, Mode mode) {
+                this.path = path;
+                this.mode = mode;
+            }
+
+            @Override
+            public Stream<Path> resolvePaths(PathLookup pathLookup) {
+                return Stream.of(path);
+            }
+
+            @Override
+            public Mode mode() {
+                return mode;
+            }
+
+            @Override
+            public boolean equals(Object obj) {
+                if (obj == this) return true;
+                if (obj == null || obj.getClass() != this.getClass()) return false;
+                var that = (AbsolutePathFileData) obj;
+                return Objects.equals(this.path, that.path) && Objects.equals(this.mode, that.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(path, mode);
+            }
+        }
+
+        final class RelativePathFileData implements FileData {
+            private final Path relativePath;
+            private final BaseDir baseDir;
+            private final Mode mode;
+
+            private RelativePathFileData(Path relativePath, BaseDir baseDir, Mode mode) {
+                this.relativePath = relativePath;
+                this.baseDir = baseDir;
+                this.mode = mode;
+            }
+
+            @Override
+            public Stream<Path> resolvePaths(PathLookup pathLookup) {
+                Objects.requireNonNull(pathLookup);
+                switch (baseDir) {
+                    case CONFIG:
+                        return Stream.of(pathLookup.configDir().resolve(relativePath));
+                    case DATA:
+                        return Arrays.stream(pathLookup.dataDirs()).map(d -> d.resolve(relativePath));
+                    default:
```

Comment:
nit: the default case shouldn't be necessary since the cases above are exhaustive?

Reply:
That is there only to protect us in case we add another field to the enum and forget to update this. 
I can use a switch expression, but that would make the backports fail IIRC (8.x language level does not support them?)